File: archon/internal/dns/cloudflare/cloudflare.go
package cloudflare

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/BlueBeard63/archon/internal/models"
)

const cloudflareAPIBase = "https://api.cloudflare.com/client/v4"

// Provider implements dns.Provider for Cloudflare
type Provider struct {
	apiToken string
	zoneID   string
	client   *http.Client
}

// NewCloudflareProvider creates a new Cloudflare DNS provider
func NewCloudflareProvider(apiToken, zoneID string) *Provider {
	return &Provider{
		apiToken: apiToken,
		zoneID:   zoneID,
		client: &http.Client{
			Timeout: 15 * time.Second,
		},
	}
}

// ListRecords retrieves all DNS records for a zone
func (p *Provider) ListRecords(domain string) ([]models.DnsRecord, error) {
	url := fmt.Sprintf("%s/zones/%s/dns_records", cloudflareAPIBase, p.zoneID)

	// Create HTTP request
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+p.apiToken)
	req.Header.Set("Content-Type", "application/json")

	// Make request
	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	// Parse response
	var cfResp struct {
		Success bool               `json:"success"`
		Errors  []cloudflareError  `json:"errors"`
		Result  []cloudflareRecord `json:"result"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&cfResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	if !cfResp.Success {
		if len(cfResp.Errors) > 0 {
			return nil, fmt.Errorf("cloudflare API error: %s", cfResp.Errors[0].Message)
		}
		return nil, fmt.Errorf("cloudflare API request failed")
	}

	// Convert to models.DnsRecord
	records := make([]models.DnsRecord, 0, len(cfResp.Result))
	for _, cfRecord := range cfResp.Result {
		records = append(records, fromCloudflareRecord(cfRecord))
	}

	return records, nil
}

// CreateRecord creates a new DNS record in Cloudflare
func (p *Provider) CreateRecord(domain string, record *models.DnsRecord, tags []string) (*models.DnsRecord, error) {
	url := fmt.Sprintf("%s/zones/%s/dns_records", cloudflareAPIBase, p.zoneID)

	// Convert to Cloudflare format
	cfRecord := toCloudflareRecord(record, tags)
	// Marshal request body
	body, err := json.Marshal(cfRecord)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Create HTTP request with body
	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+p.apiToken)
	req.Header.Set("Content-Type", "application/json")

	// Make request
	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	// Parse response
	var cfResp cloudflareResponse
	if err := json.NewDecoder(resp.Body).Decode(&cfResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	if !cfResp.Success {
		if len(cfResp.Errors) > 0 {
			return nil, fmt.Errorf("cloudflare API error: %s", cfResp.Errors[0].Message)
		}
		return nil, fmt.Errorf("cloudflare API request failed")
	}

	// Parse the created record from response
	var createdRecord cloudflareRecord
	if err := json.Unmarshal(cfResp.Result, &createdRecord); err != nil {
		return nil, fmt.Errorf("failed to parse created record: %w", err)
	}

	// Convert back to models.DnsRecord
	resultRecord := fromCloudflareRecord(createdRecord)
	return &resultRecord, nil
}

// UpdateRecord updates an existing DNS record
func (p *Provider) UpdateRecord(domain string, record *models.DnsRecord, tags []string) (*models.DnsRecord, error) {
	if record.ID == nil || *record.ID == "" {
		return nil, fmt.Errorf("record ID is required for updates")
	}

	url := fmt.Sprintf("%s/zones/%s/dns_records/%s", cloudflareAPIBase, p.zoneID, *record.ID)

	// Convert to Cloudflare format
	cfRecord := toCloudflareRecord(record, tags)

	// Marshal request body
	body, err := json.Marshal(cfRecord)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Create HTTP request
	req, err := http.NewRequest("PUT", url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+p.apiToken)
	req.Header.Set("Content-Type", "application/json")

	// Make request
	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	// Parse response
	var cfResp cloudflareResponse
	if err := json.NewDecoder(resp.Body).Decode(&cfResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	if !cfResp.Success {
		if len(cfResp.Errors) > 0 {
			return nil, fmt.Errorf("cloudflare API error: %s", cfResp.Errors[0].Message)
		}
		return nil, fmt.Errorf("cloudflare API request failed")
	}

	// Parse the updated record from response
	var updatedRecord cloudflareRecord
	if err := json.Unmarshal(cfResp.Result, &updatedRecord); err != nil {
		return nil, fmt.Errorf("failed to parse updated record: %w", err)
	}

	// Convert back to models.DnsRecord
	resultRecord := fromCloudflareRecord(updatedRecord)
	return &resultRecord, nil
}

// DeleteRecord removes a DNS record from Cloudflare
func (p *Provider) DeleteRecord(domain string, recordID string) error {
	if recordID == "" {
		return fmt.Errorf("record ID is required for deletion")
	}

	url := fmt.Sprintf("%s/zones/%s/dns_records/%s", cloudflareAPIBase, p.zoneID, recordID)

	// Create HTTP request
	req, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+p.apiToken)
	req.Header.Set("Content-Type", "application/json")

	// Make request
	resp, err := p.client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	// Parse response
	var cfResp cloudflareResponse
	if err := json.NewDecoder(resp.Body).Decode(&cfResp); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if !cfResp.Success {
		if len(cfResp.Errors) > 0 {
			return fmt.Errorf("cloudflare API error: %s", cfResp.Errors[0].Message)
		}
		return fmt.Errorf("cloudflare API request failed")
	}

	return nil
}

// cloudflareResponse represents the standard Cloudflare API response format
type cloudflareResponse struct {
	Success bool              `json:"success"`
	Errors  []cloudflareError `json:"errors"`
	Result  json.RawMessage   `json:"result"`
}

type cloudflareError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// cloudflareRecord represents a DNS record in Cloudflare's format
type cloudflareRecord struct {
	ID      string `json:"id,omitempty"`
	Type    string `json:"type"`
	Name    string `json:"name"`
	Content string `json:"content"`
	TTL     int    `json:"ttl"`
	Proxied bool   `json:"proxied"`
	Comment string `json:"comment,omitempty"`
}

// toCloudflareRecord converts models.DnsRecord to Cloudflare format
func toCloudflareRecord(record *models.DnsRecord, tags []string) cloudflareRecord {
	cfRec := cloudflareRecord{
		Type:    string(record.RecordType),
		Name:    record.Name,
		Content: record.Value,
		TTL:     record.TTL,
		Proxied: record.Proxied,
	}

	// Use first tag as comment if tags are provided
	if len(tags) > 0 {
		cfRec.Comment = tags[0]
	}

	// Include ID if it exists (for updates)
	if record.ID != nil {
		cfRec.ID = *record.ID
	}

	// Set default TTL if not specified
	if cfRec.TTL == 0 {
		cfRec.TTL = 300 // 5 minutes default
	}

	return cfRec
}

// fromCloudflareRecord converts Cloudflare format to models.DnsRecord
func fromCloudflareRecord(cf cloudflareRecord) models.DnsRecord {
	id := cf.ID
	return models.DnsRecord{
		ID:         &id,
		RecordType: models.DnsRecordType(cf.Type),
		Name:       cf.Name,
		Value:      cf.Content,
		TTL:        cf.TTL,
		Proxied:    cf.Proxied,
	}
}

File: client/src/dns/cloudflare.rs
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
use serde::{Deserialize, Serialize};
use std::time::Duration;
use thiserror::Error;

use crate::models::{DnsRecord, DnsRecordType};

const CLOUDFLARE_API_BASE: &str = "https://api.cloudflare.com/client/v4";

// ---------------------------------------------------------------------------
// Error type
// ---------------------------------------------------------------------------

#[derive(Debug, Error)]
pub enum DnsError {
    #[error("HTTP error: {0}")]
    Http(#[from] reqwest::Error),
    #[error("Cloudflare API error: {0}")]
    CloudflareApi(String),
    #[error("record ID is required for {0}")]
    MissingRecordId(&'static str),
    #[error("{0}")]
    Other(String),
}

// ---------------------------------------------------------------------------
// Cloudflare wire types
// ---------------------------------------------------------------------------

#[derive(Debug, Deserialize)]
struct CloudflareResponse {
    success: bool,
    errors: Vec<CloudflareError>,
    result: serde_json::Value,
}

#[derive(Debug, Deserialize)]
struct CloudflareError {
    #[allow(dead_code)]
    code: i32,
    message: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct CloudflareRecord {
    #[serde(skip_serializing_if = "Option::is_none")]
    id: Option<String>,
    #[serde(rename = "type")]
    record_type: String,
    name: String,
    content: String,
    ttl: i32,
    proxied: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    comment: Option<String>,
}

fn to_cloudflare_record(record: &DnsRecord, tags: &[String]) -> CloudflareRecord {
    let mut cf = CloudflareRecord {
        id: record.id.clone(),
        record_type: record.record_type.to_string(),
        name: record.name.clone(),
        content: record.value.clone(),
        ttl: if record.ttl == 0 { 300 } else { record.ttl },
        proxied: record.proxied,
        comment: tags.first().cloned(),
    };
    // Don't send ID on create
    if cf.id.as_deref() == Some("") {
        cf.id = None;
    }
    cf
}

fn from_cloudflare_record(cf: CloudflareRecord) -> DnsRecord {
    let record_type = match cf.record_type.as_str() {
        "A" => DnsRecordType::A,
        "AAAA" => DnsRecordType::AAAA,
        "CNAME" => DnsRecordType::CNAME,
        "MX" => DnsRecordType::MX,
        "TXT" => DnsRecordType::TXT,
        "SRV" => DnsRecordType::SRV,
        _ => DnsRecordType::A,
    };

    DnsRecord {
        id: cf.id,
        record_type,
        name: cf.name,
        value: cf.content,
        ttl: cf.ttl,
        proxied: cf.proxied,
    }
}

// ---------------------------------------------------------------------------
// DnsProvider trait
// ---------------------------------------------------------------------------

#[allow(async_fn_in_trait)]
pub trait DnsProvider: Send + Sync {
    async fn list_records(&self, domain: &str) -> Result<Vec<DnsRecord>, DnsError>;
    async fn create_record(
        &self,
        domain: &str,
        record: &DnsRecord,
        tags: &[String],
    ) -> Result<DnsRecord, DnsError>;
    async fn update_record(
        &self,
        domain: &str,
        record: &DnsRecord,
        tags: &[String],
    ) -> Result<DnsRecord, DnsError>;
    async fn delete_record(&self, domain: &str, record_id: &str) -> Result<(), DnsError>;
}

// ---------------------------------------------------------------------------
// CloudflareProvider
// ---------------------------------------------------------------------------

pub struct CloudflareProvider {
    api_token: String,
    zone_id: String,
    client: reqwest::Client,
}

impl CloudflareProvider {
    pub fn new(api_token: String, zone_id: String) -> Self {
        let client = reqwest::Client::builder()
            .timeout(Duration::from_secs(15))
            .build()
            .expect("failed to build HTTP client");

        Self {
            api_token,
            zone_id,
            client,
        }
    }

    fn headers(&self) -> HeaderMap {
        let mut headers = HeaderMap::new();
        headers.insert(
            AUTHORIZATION,
            HeaderValue::from_str(&format!("Bearer {}", self.api_token))
                .expect("invalid API token"),
        );
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
        headers
    }

    fn check_response(resp: &CloudflareResponse) -> Result<(), DnsError> {
        if !resp.success {
            let msg = resp
                .errors
                .first()
                .map(|e| e.message.clone())
                .unwrap_or_else(|| "cloudflare API request failed".into());
            return Err(DnsError::CloudflareApi(msg));
        }
        Ok(())
    }
}

impl DnsProvider for CloudflareProvider {
    async fn list_records(&self, _domain: &str) -> Result<Vec<DnsRecord>, DnsError> {
        let url = format!("{}/zones/{}/dns_records", CLOUDFLARE_API_BASE, self.zone_id);

        let resp = self
            .client
            .get(&url)
            .headers(self.headers())
            .send()
            .await?;

        let cf_resp: CloudflareResponse = resp.json().await?;
        Self::check_response(&cf_resp)?;

        let cf_records: Vec<CloudflareRecord> =
            serde_json::from_value(cf_resp.result).map_err(|e| DnsError::Other(e.to_string()))?;

        Ok(cf_records.into_iter().map(from_cloudflare_record).collect())
    }

    async fn create_record(
        &self,
        _domain: &str,
        record: &DnsRecord,
        tags: &[String],
    ) -> Result<DnsRecord, DnsError> {
        let url = format!("{}/zones/{}/dns_records", CLOUDFLARE_API_BASE, self.zone_id);
        let cf_record = to_cloudflare_record(record, tags);

        let resp = self
            .client
            .post(&url)
            .headers(self.headers())
            .json(&cf_record)
            .send()
            .await?;

        let cf_resp: CloudflareResponse = resp.json().await?;
        Self::check_response(&cf_resp)?;

        let created: CloudflareRecord =
            serde_json::from_value(cf_resp.result).map_err(|e| DnsError::Other(e.to_string()))?;

        Ok(from_cloudflare_record(created))
    }

    async fn update_record(
        &self,
        _domain: &str,
        record: &DnsRecord,
        tags: &[String],
    ) -> Result<DnsRecord, DnsError> {
        let record_id = record
            .id
            .as_deref()
            .filter(|id| !id.is_empty())
            .ok_or(DnsError::MissingRecordId("updates"))?;

        let url = format!(
            "{}/zones/{}/dns_records/{}",
            CLOUDFLARE_API_BASE, self.zone_id, record_id
        );
        let cf_record = to_cloudflare_record(record, tags);

        let resp = self
            .client
            .put(&url)
            .headers(self.headers())
            .json(&cf_record)
            .send()
            .await?;

        let cf_resp: CloudflareResponse = resp.json().await?;
        Self::check_response(&cf_resp)?;

        let updated: CloudflareRecord =
            serde_json::from_value(cf_resp.result).map_err(|e| DnsError::Other(e.to_string()))?;

        Ok(from_cloudflare_record(updated))
    }

    async fn delete_record(&self, _domain: &str, record_id: &str) -> Result<(), DnsError> {
        if record_id.is_empty() {
            return Err(DnsError::MissingRecordId("deletion"));
        }

        let url = format!(
            "{}/zones/{}/dns_records/{}",
            CLOUDFLARE_API_BASE, self.zone_id, record_id
        );

        let resp = self
            .client
            .delete(&url)
            .headers(self.headers())
            .send()
            .await?;

        let cf_resp: CloudflareResponse = resp.json().await?;
        Self::check_response(&cf_resp)?;

        Ok(())
    }
}

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

use crate::models::{DnsProvider as DnsProviderModel, DnsProviderType};

/// Creates a CloudflareProvider from a DnsProvider model config.
/// Returns None for Manual DNS, Err for Route53 (unimplemented).
pub fn create_provider(
    provider: &DnsProviderModel,
) -> Result<Option<CloudflareProvider>, DnsError> {
    match provider.provider_type {
        DnsProviderType::Cloudflare => {
            let token = provider
                .api_token
                .as_deref()
                .filter(|t| !t.is_empty())
                .ok_or_else(|| {
                    DnsError::Other("Cloudflare provider requires api_token".into())
                })?;
            let zone_id = provider
                .zone_id
                .as_deref()
                .filter(|z| !z.is_empty())
                .ok_or_else(|| {
                    DnsError::Other("Cloudflare provider requires zone_id".into())
                })?;
            Ok(Some(CloudflareProvider::new(
                token.to_string(),
                zone_id.to_string(),
            )))
        }
        DnsProviderType::Route53 => Err(DnsError::Other(
            "Route53 provider not yet implemented".into(),
        )),
        DnsProviderType::Manual => Ok(None),
    }
}
